### DISCUSSION TITLE
What’s Ram drive? and how to fix it?

### DISCUSSION THREAD

**Discussion by MueJosh** [Nachtzuster/BirdNET-Pi#25] [General]
Pi4 4gb 

I've just tested the experimental ram drive -> tools -> services 

At first everything was working fine (I let it run for 10 minutes) but didn't want to test it any further. 
After I deactivated it again the problems showed up

(Yeah it says experimental- I know^^)

After bootlooping for 10 minutes it's working fine but then it crashes (I can't access the pi in any way e.g. website, ssh) 
Its not that problemativ since ive got 4 running birdnet setups right now at my place

Is there a way to undo my mistake? 
Or do i need a fresh install? (Not problematic- it just takes time) 

I still wonder what "RAM Drive" is- less sd card wear? 
Whats the minimum ram needed? Is the pi 02w fine for that?
I dont activate it again any time soon tho ;)

EDIT: after rebooting for the xx'th time, it worked- i hitted update in the tools bar and after that it's now running without issues.

**Comment by Nachtzuster** [comment]
Well, I'm glad it ran fine when you enabled it :-)

When you enable/disable it, the RPi will reboot, but it is not very obvious on the ui when it is back and running again, unfortunately (PR's are welcome !).
Without logs it hard to tell what was going on.

Anyway, when it is `inactive` in the ui, then that should be that: it is off.

Technically it is meant to reduce sd card wear writing all transient files to it, but you need plenty of RAM, probably min 2G-4G.
More in depth: it mounts tmpfs on top of $HOME/BirdSongs/StreamData/

**Reply by arne1921KF** [reply]
I'm testing the feature for the first time today. So far, it works great. 

Asked for clarification - I am fairly new to the linux world and RasPi's: 

The recordings in `/BirdSongs/StreamData` are not stored on the SD, but in a temporary file system (aka RAM disk) mounted as this directory, correct? 

When processing of those 'streamed' recordings in three-second chunks is finished and a) the detection threshold was met, and b) the privacy threshold is not met, a recording (of the length set in the variable `$RECORDING_LENGTH`) is saved to `/BirdSongs/Extracted`, which is not a [tmpfs](https://en.wikipedia.org/wiki/Tmpfs), but resides on the mounted SD.  

Is my understanding correct? There is no persistent recording of the audio input? Since the option `noatime` is set for the root directory, and `relatime` for the tmpfs, I think there should also be no persistent metadata for the streamed recordings unless written to the SD, correct?

Would you mind to explain why you deem the feature experimental? The risk I see so far are outweigh by the gains.
One risk: the tpfs might become a memory hog - if many birds are around and the Pi cannot process the data fast enough. I am running it on a Raspi 4B with 4 GB, and so far it does not look like this is going to be an issue on a "dedicated" BirdNet-Pi, ever.
A second risk: loss of recent recordings on power loss. This will happen, of course. If this does not affect system stability, this is a price well worth paying.

**Reply by Nachtzuster** [reply]
Your explanation is pretty much correct.

> Since the option `noatime` is set for the root directory, and `relatime` for the tmpfs, I think there should also be no persistent metadata for the streamed recordings unless written to the SD, correct?

This part I don't quite understand, so I can't comment

It is experimental, because the two reasons you point out:
- it can become a memory hog when it processing gets behind (maybe due to notifications to Birdweather, apprise)
- loss of recent recordings on power loss or rebooting by the user

Also, 2024 SD cards seem to be far more robust than, say, 5-10 years ago, (but this is hard to quantify), so I'm not sure it is really worth it.

**Reply by arne1921KF** [reply]
Sorry for the delayed response. 

To elaborate on my [atime](https://tldp.org/LDP/solrhe/Securing-Optimizing-Linux-RH-Edition-v1.3/chap6sec73.html) question:

My understanding is that this option prevents that the access time for a file is written. My understanding is also that this is set for the root partition, and hence access to files on the SD will not be logged. 

If that is correct, even a forensic hack should not be able to identify the time at which something which is *not* written to the database was detected. I could not even reconstruct the time anything *other* than targeted has been detected (and the result with the accompanying sound file has *not* been written to the database), even if operations for that detection needed access to a file on the SD. 

This is a very complicated line of thought, but in matters of privacy, I am used to people trying to cover all bases. 😬

**Comment by MueJosh** [comment]
> Well, I'm glad it ran fine when you enabled it :-)

Sounds like it could've destroy my system - lol 

What's PR? 
The "view log"-output looked really different, that's how I knew it was running. 

Alright, I'll test it again in a few days (right now I test some others birdnet applications and scripts and need a stable reference) 


In my whole lifetime I was never able to kill a sd card (I really shouldn't write that...) but the idea seems great- especially for systems without physical access. 

Pi 4b 4gb for home usage 


I'll check it out 👍 
Thank you for the explanation
